### PR TITLE
Refactor decoupled clock to better handle negative seeks

### PR DESCRIPTION
**Pull request by smoogipoo**
`DecoupledInterpolatingFramedClock` couldn't handle negative seeks at all.

During `Seek()`, it stops the source clock and calls `ProcessFrame` which attempts to start the source clock. This only happens if the seek on the source clock to `CurrentTime` succeeds.  

The problem is that the source clock is seeked to the `CurrentTime` from the _previous_ frame (i.e. before the `Seek()`), which would always succeed unless `CurrentTime` was negative in the first place.  

It would then follow through and use the interpolated source time, which would be the time from the last frame.

### CODE REVIEW

File: osu.Framework/Timing/DecoupleableInterpolatingFramedClock.cs [hunk 1]
```diff
@@ -62,40 +64,46 @@ public override void ProcessFrame()
...
 
-            double proposedTime = useInterpolatedSourceTime ? base.CurrentTime : decoupledClock.CurrentTime;
+                // since the decoupled clock was started, update it once more.
+                decoupledClock.ProcessFrame();
+            }
 
             elapsedFrameTime = useInterpolatedSourceTime ? base.ElapsedFrameTime : decoupledClock.ElapsedFrameTime;
-
-            currentTime = elapsedFrameTime < 0 ? proposedTime : Math.Max(currentTime, proposedTime);
+            currentTime = elapsedFrameTime < 0 ? ProposedCurrentTime : Math.Max(currentTime, proposedTime);
```
Comment (peppy) [inline osu.Framework/Timing/DecoupleableInterpolatingFramedClock.cs]: the only thing i'm not sure about is the reasoning behind using `ProposedCurrentTime` when negative here

Comment (peppy) [reply osu.Framework/Timing/DecoupleableInterpolatingFramedClock.cs]: All tests pass even after changing this to the local `proposedTime`, so gonna leave this PR until it's clarified.

Comment (smoogipoo) [reply osu.Framework/Timing/DecoupleableInterpolatingFramedClock.cs]: Ah yeah my eyes glossed over that, should be the local one, but good point there needs to be a test for this.

**Comment by smoogipoo** [comment]
Whoops, that was a bad rebase. My bad - not gonna force-push to fix it.

**Comment by smoogipoo** [comment]
I've refactored this a little bit further - haven't added any additional tests or anything, just consolidating the logic and making sense of it. All existing test cases passed and from `TestSceneClocks` it all seemed ok.

**Pull request merged**